src/main/kotlin/app/revanced/patcher/InstructionFilter.kt

```diff
@@ -0,0 +1,313 @@
+package app.revanced.patcher
+
+import app.revanced.patcher.InstructionFilter.Companion.METHOD_MAX_INSTRUCTIONS
+import app.revanced.patcher.extensions.InstructionExtensions.instructions
+import com.android.tools.smali.dexlib2.Opcode
+import com.android.tools.smali.dexlib2.iface.ClassDef
+import com.android.tools.smali.dexlib2.iface.Method
+import com.android.tools.smali.dexlib2.iface.instruction.Instruction
+import com.android.tools.smali.dexlib2.iface.instruction.ReferenceInstruction
+import com.android.tools.smali.dexlib2.iface.instruction.WideLiteralInstruction
+import com.android.tools.smali.dexlib2.iface.reference.FieldReference
+import com.android.tools.smali.dexlib2.iface.reference.MethodReference
+import java.util.EnumSet
+import kotlin.collections.forEach
+
+interface InstructionFilter {
+    /**
+     * Maximum number of non matching instructions that can be before this filter.
+     * A value of zero means this filter must match immediately after the prior filter,
+     * or if this is the first filter then this may only match the first instruction of the method.
+     */
+    val maxInstructionsBefore: Int
```

Comment:
InstructionFilter could be turned into an abstract class, and maxInstructionsBefore a constructor property